flink-connectors/flink-connector-kinesis/src/test/java/org/apache/flink/streaming/connectors/kinesis/proxy/KinesisProxyTest.java

```diff
@@ -70,20 +113,107 @@ public void testIsRecoverableExceptionWithNullErrorType() {
+						"shardId-000000000002",
+						"shardId-000000000003");
+		shardIdSet = new HashSet<>(shardIds);
+		shards =
+				shardIds
+						.stream()
+						.map(shardId -> new Shard().withShardId(shardId))
+						.collect(Collectors.toList());
+		Properties kinesisConsumerConfig = new Properties();
+		kinesisConsumerConfig.setProperty(ConsumerConfigConstants.AWS_REGION, "us-east-1");
+		kinesisConsumerConfig.setProperty(ConsumerConfigConstants.AWS_ACCESS_KEY_ID, "fake_accesskey");
+		kinesisConsumerConfig.setProperty(
+				ConsumerConfigConstants.AWS_SECRET_ACCESS_KEY, "fake_secretkey");
+		KinesisProxy kinesisProxy = new KinesisProxy(kinesisConsumerConfig);
+		AmazonKinesis mockClient = mock(AmazonKinesis.class);
+		Whitebox.setInternalState(kinesisProxy, "kinesisClient", mockClient);
+
+		ListShardsResult responseWithMoreData =
+				new ListShardsResult().withShards(shards.subList(0, 2)).withNextToken(NEXT_TOKEN);
+		ListShardsResult responseFinal =
+				new ListShardsResult().withShards(shards.subList(2, shards.size())).withNextToken(null);
+		doReturn(responseWithMoreData)
+				.when(mockClient)
+				.listShards(argThat(initialListShardsRequestMatcher()));
+		doReturn(responseFinal).when(mockClient).listShards(argThat(listShardsNextToken(NEXT_TOKEN)));
+		HashMap<String, String> streamHashMap =
+				createInitialSubscribedStreamsToLastDiscoveredShardsState(Arrays.asList(fakeStreamName));
+		GetShardListResult shardListResult = kinesisProxy.getShardList(streamHashMap);
+
+		Assert.assertEquals(shardListResult.hasRetrievedShards(), true);
+
+		Set<String> expectedStreams = new HashSet<>();
+		expectedStreams.add(fakeStreamName);
+		Assert.assertEquals(shardListResult.getStreamsWithRetrievedShards(), expectedStreams);
+		List<StreamShardHandle> actualShardList =
+				shardListResult.getRetrievedShardListOfStream(fakeStreamName);
+		List<StreamShardHandle> expectedStreamShard = new ArrayList<>();
+		System.out.println(actualShardList.toString());
+		assertThat(actualShardList, hasSize(4));
+		for (int i = 0; i < 4; i++) {
+			StreamShardHandle shardHandle =
+					new StreamShardHandle(
+							fakeStreamName,
+							new Shard().withShardId(KinesisShardIdGenerator.generateFromShardOrder(i)));
+			expectedStreamShard.add(shardHandle);
+		}
+
+		Assert.assertThat(
+				actualShardList,
+				containsInAnyOrder(
+						expectedStreamShard.toArray(new StreamShardHandle[actualShardList.size()])));
+	}
+
+	private static class ListShardsRequestMatcher extends TypeSafeDiagnosingMatcher<ListShardsRequest> {
+		private final String shardId;
+		private final String nextToken;
+
+		ListShardsRequestMatcher(String shardIdArg, String nextTokenArg) {
+			shardId = shardIdArg;
+			nextToken = nextTokenArg;
+		}
+
+		@Override
+		protected boolean matchesSafely(final ListShardsRequest listShardsRequest, final Description description) {
+			if (shardId == null) {
+				if (StringUtils.isNotEmpty(listShardsRequest.getExclusiveStartShardId())) {
```

Comment:
Can we avoid using `StringUtils` and just use `!String.isEmpty()` instead?

Reply:
Done. Also note that there is quite a bit of code from reused from here: https://goo.gl/MGSRb3. Do I need to add reference to that code?